Slaughter demon bug fixes and adds laughter demon

coiax opened the pull request:
Fixes #16926.
:cl: coiax
rscadd: Laughter demons are like slaughter demons, but they tickle people to death
tweak: Slaughter demons are no longer able to cast spells while phased.
tweak: Wizards can now speak while ethereal, but still cannot cast spells while ethereal.
/:cl:
- [x] actually fix Mehki's issue
- [x] find out why the mobs being added to consumed_mobs are being deleted

coiax committed:
Restructure demon consumption

coiax committed:
Fixes victim mobs being deleted when eaten

This single line fix prevents the consumed mob from being deleted
because of REASONS. Now rejoyce as you recover the eaten bodies of your
friends when you finally kill the demon.

Zilenan91 commented:
It's not a bug that they gib people now, it's to prevent demons from being tracked on suit sensors

FantasticFwoosh commented:
I always thought the demons were removing people from the physical plain by moving them into the blood channel dimension. I wasnt aware the deletion was unintentional.

Zilenan91 commented:
It's not unintentional, kor added it back in

coiax commented:
There's a bunch of comments and messages in the code that suggest that slaughter demons, like morphlings, are supposed to drop their contents on death. If @KorPhaeron made this change out of concern for suit sensors, I can adjust accordingly.

lordpidey commented on code/modules/mob/living/bloodcrawl.dm:
```diff
@@ -37,58 +40,93 @@
+	if(istype(src, /mob/living/simple_animal/slaughter))
+		SD = src
+		SD.consumed_mobs += victim
+		world << "DEBUG: [victim] added to [SD] consumed mobs"
```
Probably didn't mean to include this

coiax committed:
Removed debug statement

KorPhaeron commented:
Adding it back in does things like

-Make them vulnerable to ling ressurection
-Make them trackable via suit sensors and the disk
-Make them die in a hilarious manner during the zombie event

I am sure there are others. Being dragged into another dimension and eaten seems like the sort of thing that should gib you anyway

coiax commented:
@KorPhaeron noted. Should I have the demon eject all of their inventory, because it's not "meat"? Or destroy that too? Also, I have no idea how changeling resurrection would even work, it's not like it's put in the demon's stomach, because it doesn't have one.

coiax committed:
Disabled keeping the bodies by default

But you can turn it back on if you want.

Zilenan91 commented:
The bodies are put in the demon's contents, like a locker, so the lings/zombies can hit it from the inside.

RemieRichards commented on code/modules/mob/living/bloodcrawl.dm:
```diff
@@ -37,58 +40,92 @@
+/mob/living/proc/bloodcrawl_consume(obj/effect/decal/cleanable/B, mob/living/victim)
+	src << "<span class='danger'>You begin to feast on [victim]. You can not move while you are doing this.</span>"
+
+	for(var/i = 3; i > 0; i--)
```
```for(var/i = 3 to 1 step -1)```

coiax replied:
literally not my code, just copying what was there before. Given all it does is loop the sound code 3 times, I'll just have a normal for loop counting up

RemieRichards replied:
then ```for(var/i in 1 to 3)```

coiax committed:
Laughter demons and bug fixes

coiax committed:
Laughter demons have bows now

coiax committed:
Fixes some sound related bugs

coiax committed:
Phasing and comments

coiax committed:
Removed etheral mute, bloodcrawl while phased

KazeEspada commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -54,21 +58,42 @@
+// icon_state, and it'll be overriden if just defined above.
+/obj/effect/decal/cleanable/blood/innards/New()
+	..()
+	icon_state = "innards"
```
You can set the random states for your innards to be list"innards")

RemieRichards replied:
yes this ^

RemieRichards commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -54,21 +58,42 @@
 	ghostize()
 	qdel(src)
+
 	return
```
trailing return, remove it please.

RemieRichards commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -54,21 +58,42 @@
+/mob/living/simple_animal/slaughter/proc/death_mealspill()
+	if(consumed_mobs)
+		for(var/mob/living/M in consumed_mobs)
+			M.loc = get_turf(src)
```
all ```thing.loc = X``` should be ```thing.forceMove(X)```

RemieRichards commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -115,3 +140,55 @@
+	// of kittens
+	var/mob/living/simple_animal/pet/cat/kitten/K = new(get_turf(src))
+	K.name = "Laughter"
+	return
```
trailing return, remove it please.

RemieRichards commented on code/modules/spells/spell.dm:
```diff
@@ -91,6 +92,11 @@ var/list/spells = typesof(/obj/effect/proc_holder/spell) //needed for the badmin
 		return 0
 
+	var/loc = user.loc
+	if(!phase_allowed && (istype(loc, /obj/effect/dummy/spell_jaunt) || istype(loc, /obj/effect/dummy/slaughter)))
```
```istype(loc, /obj/effect/dummy)```

RemieRichards commented on code/modules/spells/spell_types/ethereal_jaunt.dm:
```diff
@@ -43,9 +43,10 @@
 			jaunt_steam(mobloc)
 
-			mute(target)
+			// Used to mute the wizard here, but adding the phased_allowed
```
don't do this, don't embed history in comments, git tracks history.

KazeEspada commented:
Heavily biased towards naming them devils. Mosly because of D&D but also because it makes them named different than slaughter demons.

";DEVIL IN THE LAWYER OFFICE!"

coiax commented:
@KazeEspada This is the wrong PR for that comment, Kaze.

coiax committed:
Coding review fixes

- Removes trailing returns
- Removes mentioning code history in comments
- Uses built in features for cleanable decals
- Uses /obj/effect/dummy class to detect for incorporeal nature

coiax commented:
Anyway, this is complete, ready for review/merge.

KazeEspada commented:
I deleted that comment and moved it to the right PR.

ChangelingRain commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -115,3 +133,54 @@
+	// them so much, it wants to hug everyone at once!
+	name = "laughter demon"
+	real_name = "laughter demon"
+	desc = "A large, adorable creature covered in armored pink scales."
```
this desc is not accurate it just has bows on its horns

coiax committed:
Corrected description of laughter demon

ChangelingRain commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -54,21 +58,35 @@
-	visible_message("<span class='danger'>[src] screams in anger as it collapses into a puddle of viscera, its most recent meals spilling out of it.</span>")
-	for(var/mob/living/M in consumed_mobs)
-		M.loc = get_turf(src)
+	death_gibs()
```
instead of this, put the stuff you want in the demon's loot var eg; `loot = list(/mob/living/simple_animal/pet/cat/kitten{name = "Laughter"}`
You don't need to set it on runtime either, you can set it as a var.

ChangelingRain commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -54,21 +58,35 @@
+	death_mealspill()
+
+	playsound(get_turf(src),death_sound, 200, 1)
+	death_message()
```
Instead of this, set the deathmessage var to what you want it to be in New()

ChangelingRain commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -54,21 +58,35 @@
 
 /mob/living/simple_animal/slaughter/death()
 	..(1)
```
Remove the `1` and instead set the del_on_death var to 1 for these.

coiax replied:
do  you mean just call super() with no arguments?

ChangelingRain replied:
ye

ChangelingRain commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -54,21 +58,35 @@
+	death_message()
+
 	ghostize()
 	qdel(src)
```
If you set the del_on_death var to 1 these are both unneeded.

ChangelingRain commented on code/modules/spells/spell_types/ethereal_jaunt.dm:
```diff
@@ -72,27 +70,7 @@
 				target.canmove = 1
 
 //Silence wizard during jaunt so they cannot spell cast while invisible
```
Remove the now incorrect comment.

coiax committed:
Code review

coiax committed:
Death message

ChangelingRain commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -57,22 +64,12 @@
 
 	death_mealspill()
 
+	visible_message(deathmessage)
```
..() already does this.

ChangelingRain commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -42,8 +42,15 @@
+				/obj/effect/decal/cleanable/blood/innards, \
+				/obj/item/organ/internal/heart/demon)
+	del_on_death = 1
+	var/death_message = "<span class='danger'>The generic slaughter demon screams in anger because this is the default death message!</span>"
```
Not var/death_message, deathmessage
It's an already-defined var(and includes the danger span class)

coiax committed:
Deathmessage changes

phil235 commented on code/modules/mob/living/bloodcrawl.dm:
```diff
@@ -37,58 +40,111 @@
+
+	var/del_corpse = TRUE
+	var/mob/living/simple_animal/slaughter/SD
+	if(istype(src, /mob/living/simple_animal/slaughter))
```
no. You override the proc so laughter and slaughter demons each have their own version of this proc. This way you won't need to use this silly consumed_mobs = null trick.

coiax replied:
That check isn't there because of slaughter demons vs laughter demons. It's there because non-slaughter demons could gain BLOODCRAWL_EAT. And they won't have a consumed_mobs var.

phil235 replied:
```
/mob/living/proc/bloodcrawl_consume(obj/effect/decal/cleanable/B, mob/living/victim, del_corpse = 1)
	stuff

/mob/living/simple_animal/slaughter/laughter/bloodcrawl_consume(obj/effect/decal/cleanable/B, mob/living/victim, del_corpse = 1)
	consumed_mobs += victim
	. = ..(B, victim, 0) //so the victim's corpse isn't deleted

/mob/living/simple_animal/slaughter/laughter/bloodcrawl_consume(obj/effect/decal/cleanable/B, mob/living/victim, del_corpse = 1)
```

phil235 commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -115,3 +126,51 @@
+
+/mob/living/simple_animal/slaughter/laughter/death_mealspill()
+	if(!consumed_mobs)
+		return
```
why check this when the list is already created for laughter demons?

coiax replied:
It could be null for REASONS. Maybe badminry to make a lethal laughter demon.

phil235 commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -32,14 +34,23 @@
-	var/list/consumed_mobs = list()
+	// Set consumed_mobs to a list() to keep the corpses in the demon
+	// for later retrieval on demon death
+	var/list/consumed_mobs = null
```
I'm really not fond of this trick to be honest.

phil235 commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -53,22 +64,22 @@
-	ghostize()
-	qdel(src)
-	return
+	..()
```
not ..(1) ?

phil235 commented on code/game/gamemodes/miniantags/slaughter/slaughter.dm:
```diff
@@ -53,22 +64,22 @@
+
+	playsound(get_turf(src),death_sound, 200, 1)
 
+/mob/living/simple_animal/slaughter/proc/death_mealspill()
```
why not put this code in /Destroy() since you use del_on_death?

coiax replied:
I'm pretty sure having side effects when calling Destroy() is a bad thing, given it led to the weird container issues we're having atm

phil235 commented on code/modules/mob/living/bloodcrawl.dm:
```diff
@@ -37,58 +40,111 @@
+	if(pulledby)
+		pulledby.stop_pulling()
+	src.loc = holder
+	src.holder = holder
```
I know it's not your code, but this is ugly. Actually the whole proc is.

phil235 commented on code/modules/mob/living/bloodcrawl.dm:
```diff
@@ -37,58 +40,111 @@
+	else if(victim.reagents && victim.reagents.has_reagent("demonsblood"))
+		visible_message("<span class='warning'>Something prevents [victim] from entering the pool!</span>", "<span class='warning'>A strange force is blocking [victim] from entering!</span>", "<span class='notice'>You hear a splash and a thud.</span>")
+	else
+		victim.loc = src
```
yeah, let's totally change the mob's loc without unbuckling him... *sigh*
The proc needs to use forceMove() for both the demon and its victims.

phil235 commented on code/modules/spells/spell.dm:
```diff
@@ -91,6 +92,11 @@ var/list/spells = typesof(/obj/effect/proc_holder/spell) //needed for the badmin
 		user << "<span class='notice'>Not when you're incapacitated.</span>"
 		return 0
 
+	var/loc = user.loc
```
terrible variable name....

phil235 commented:
Mixing bugfixes and new features in the same PR is against our policy (https://github.com/tgstation/-tg-station/blob/master/CONTRIBUTING.md).

coiax commented:
I don't think that's a particularly realistic standard to uphold tbh.

coiax committed:
Code review

coiax committed:
Merge branch 'master' into slaughter-fixes

coiax committed:
Refactoring and tweaks
